src/Analyzers/CSharp/Analyzers/FileBasedPrograms/AppDirectiveHelpers.cs

```diff
@@ -0,0 +1,705 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// Largely copied from https://github.com/dotnet/sdk/blob/ce691a53dad7f2b6057c85c1e7fd7800dfe0788f/src/Cli/dotnet/Commands/Run/VirtualProjectBuildingCommand.cs#L1410
+// TODO: we'd like to extract a source package which makes this copy of the source unnecessary.
```

Comment:
This looks good to me overall. Do you intend to create the source package yourself in the sdk, or should I go do and prioritize that?